### PR TITLE
Added webspace settings

### PR DESCRIPTION
**Pull request by wachterjohannes**
__tasks:__

- [x] test coverage
- [ ] submit changes to the documentation
- [x] Javascript bootstrap
- [x] WebspaceController
- [x] Default snippet per type
  * extract save and load default from controller
  * translate uuid into title of snippet (user-locale)
- [x] General informations
- [x] Main URL
- [x] Sorting navigation and content navigation
- [x] table-view nested improvements
- [x] change route to: `content/webspace/settings:sulu_io/general`

__informations:__

| q                | a
| ---------------- | ---
| Fixed tickets    | none
| Related PRs      | https://github.com/massiveart/husky/pull/586 https://github.com/sulu-io/sulu-standard/pull/569
| BC breaks        | none
| Documentation PR | TODO

### CODE REVIEW

File: src/Sulu/Bundle/AdminBundle/Tests/Unit/Navigation/ContentNavigationRegistryTest.php [hunk 1]
```diff
@@ -98,12 +115,35 @@ public function testGetNavigationItems($mappings, $options, $contentNavigationDa
 
         foreach ($contentNavigationData as $service => $items) {
             $contentNavigationProvider = $this->prophesize(ContentNavigationProviderInterface::class);
+
+            $items = array_map(
+                function ($item) {
+
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Tests/Unit/Navigation/ContentNavigationRegistryTest.php]: No new line here

File: src/Sulu/Bundle/ContentBundle/Controller/WebspaceController.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+namespace Sulu\Bundle\ContentBundle\Controller;
+
+use FOS\RestBundle\Routing\ClassResourceInterface;
+use Sulu\Bundle\ContentBundle\Admin\ContentAdmin;
+use Sulu\Component\Rest\RestController;
+use Sulu\Component\Security\SecuredControllerInterface;
+
+class WebspaceController extends RestController implements ClassResourceInterface, SecuredControllerInterface
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Controller/WebspaceController.php]: Missing PHPDoc

File: src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+use JMS\Serializer\EventDispatcher\Events;
+use JMS\Serializer\EventDispatcher\EventSubscriberInterface;
+use JMS\Serializer\EventDispatcher\ObjectEvent;
+use JMS\Serializer\JsonSerializationVisitor;
+use Sulu\Component\Webspace\Manager\WebspaceManagerInterface;
+use Sulu\Component\Webspace\PortalInformation;
+use Sulu\Component\Webspace\Webspace;
+
+class WebspaceSerializeEventSubscriber implements EventSubscriberInterface
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php]: Missing PHPDoc

File: src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        $visitor = $event->getVisitor();
+
+        if (!($webspace instanceof Webspace)) {
+            return;
+        }
+
+        $portalInformation = $this->webspaceManager->getPortalInformations($this->environment);
+
+        $portalInformation = $event->getContext()->accept(
+            array_filter(
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php]: Wouldn't it be more efficient and cleaner to do stuff like this in the `WebspaceManager`?

File: src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                    return $information->getWebspaceKey() === $webspace->getKey();
+                }
+            )
+        );
+
+        // FIXME dirty hack to avoid same "id" for datagrid
+        $i = 0;
+        $portalInformation = array_map(
+            function ($item) use (&$i) {
+                $item['id'] = ++$i;
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php]: Isn't it possible to use the unique key of the webspace instead a number? Then we would not need this hack.

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/webspace/settings/analytics/main.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+define([], function() {
+
+    'use strict';
+
+    var defaults = {};
+
+    return {
+
+        defaults: defaults,
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/webspace/settings/analytics/main.js]: Do you want to do this, even if the defaults are empty?

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/webspace/settings/main.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            ).then(function(data) {
+                deferred.resolve(data);
+            });
+
+            return deferred.promise();
+        },
+
+        initialize: function() {
+            this.sandbox.on('sulu.header.back', function() {
+                this.sandbox.logger.log('TODO back');
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/webspace/settings/main.js]: Can't you implement that already?

**Comment by danrot** [comment]
Don't you want to only include the tabs, which are already done? Otherwise things might be overseen when reviewing the next PR.

File: src/Sulu/Bundle/ContentBundle/Resources/public/scss/modules/webspace-settings.scss [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        color: $grayLight;
+        white-space: nowrap;
+    }
+
+    p.value {
+        color: $grayLight;
+        font-size: $baseFontSize;
+
+        &.big {
+            font-size: $reallyBigFontSize;
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/public/scss/modules/webspace-settings.scss]: Was this variable name already existing before?

File: src/Sulu/Bundle/SnippetBundle/Admin/WebspaceContentNavigationProvider.php [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getNavigationItems(array $options = [])
+    {
+        $snippets = new ContentNavigationItem('content-navigation.webspace.snippets');
+        $snippets->setId('tab-snippets');
+        $snippets->setAction('snippets');
+        $snippets->setPosition(25);
```
Comment (danrot) [inline src/Sulu/Bundle/SnippetBundle/Admin/WebspaceContentNavigationProvider.php]: Why not continuing using factors for 10?

File: src/Sulu/Bundle/SnippetBundle/Controller/SnippettypesController.php [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+use Sulu\Component\Content\Compat\StructureManagerInterface;
+use Sulu\Component\Rest\RequestParametersTrait;
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * Handles snippet types and defaults.
+ */
+class SnippettypesController extends Controller implements ClassResourceInterface
```
Comment (danrot) [inline src/Sulu/Bundle/SnippetBundle/Controller/SnippettypesController.php]: Why did you change the name? It doesn't fit our conventions. To correct the resulting route (I guess that's what you have intended with that) we could also use the `Route` annotation, as we are doing in some other places.

File: src/Sulu/Bundle/SnippetBundle/Controller/SnippettypesController.php [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+            $template = [
+                'template' => $type->getKey(),
+                'title' => $type->getLocalizedTitle($this->getUser()->getLocale()),
+            ];
+
+            if ($defaults) {
+                $default = $defaultSnippetManager->load($webspaceKey, $type->getKey(), $this->getUser()->getLocale());
+
+                $template['defaultUuid'] = !$default ? null : $default->getUuid();
+                $template['defaultTitle'] = !$default ? null : $default->getTitle();
```
Comment (danrot) [inline src/Sulu/Bundle/SnippetBundle/Controller/SnippettypesController.php]: I would prefer positive comparisons:

```php
$default ? $default->getUuid() : null;
```

File: src/Sulu/Bundle/SnippetBundle/Controller/SnippetTypesController.php [hunk 12]
```diff
@@ -1,51 +0,0 @@
...
-            ];
-        }
-
-        $data = [
-            '_embedded' => $templates,
-            'total' => count($templates),
-        ];
-
-        return new JsonResponse($data);
-    }
```
Comment (danrot) [inline src/Sulu/Bundle/SnippetBundle/Controller/SnippetTypesController.php]: Where was this controller used?

File: src/Sulu/Bundle/SnippetBundle/Controller/SnippettypesController.php [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+            $default,
+            $this->getUser()->getLocale()
+        );
+
+        return new JsonResponse(
+            [
+                'template' => $type->getKey(),
+                'title' => $type->getLocalizedTitle($this->getUser()->getLocale()),
+                'defaultUuid' => !$defaultSnippet ? null : $defaultSnippet->getUuid(),
+                'defaultTitle' => !$defaultSnippet ? null : $defaultSnippet->getTitle(),
```
Comment (danrot) [inline src/Sulu/Bundle/SnippetBundle/Controller/SnippettypesController.php]: Same as above

File: src/Sulu/Bundle/SnippetBundle/Resources/public/js/components/snippet/form/main.js [hunk 14]
```diff
@@ -60,7 +60,7 @@ define([
                     template: {
                         options: {
                             dropdownOptions: {
-                                url: '/admin/api/snippet/types',
+                                url: '/admin/api/snippettypes',
```
Comment (danrot) [inline src/Sulu/Bundle/SnippetBundle/Resources/public/js/components/snippet/form/main.js]: I guess here the correct url would be `/admin/api/snippet-types`, maybe the FOSRestBundle can be configured this way?

File: src/Sulu/Bundle/SnippetBundle/Snippet/SnippetNotFoundException.php [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+class SnippetNotFoundException extends \Exception
+{
+    /**
+     * @var string
+     */
+    private $uuid;
+
+    public function __construct($uuid)
+    {
+        $this->uuid = $uuid;
```
Comment (danrot) [inline src/Sulu/Bundle/SnippetBundle/Snippet/SnippetNotFoundException.php]: Please add a message

File: src/Sulu/Bundle/SnippetBundle/Snippet/WrongSnippetTypeException.php [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+
+    /**
+     * @var SnippetDocument
+     */
+    private $document;
+
+    public function __construct($key, SnippetDocument $document)
+    {
+        $this->key = $key;
+        $this->document = $document;
```
Comment (danrot) [inline src/Sulu/Bundle/SnippetBundle/Snippet/WrongSnippetTypeException.php]: Please add a message

File: src/Sulu/Component/Webspace/Settings/SettingsManager.php [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    /**
+     * Returns phpcr-propertyname for given key.
+     *
+     * @param string $key
+     *
+     * @return string
+     */
+    private function propertyName($key)
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Settings/SettingsManager.php]: Verb missing, would call it something like `getPropertyName`

File: src/Sulu/Component/Webspace/Settings/SettingsManagerInterface.php [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+
+    /**
+     * Load webspace settings by wildcard.
+     *
+     * @param $webspaceKey
+     * @param $wildcard
+     *
+     * @return string[]
+     */
+    public function loadStringByWildcard($webspaceKey, $wildcard);
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Settings/SettingsManagerInterface.php]: The last two methods have the same PHPDoc, can you make clear what's the difference?

**Comment by danrot** [comment]
Code is reviewed, but not tested yet, and StyleCI fails.

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/webspace/settings/analytics/main.js (hunk 6 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/webspace/settings/analytics/main.js]: this component is only bootstraped and will be filled in one of the following PRs

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/components/webspace/settings/main.js (hunk 7 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/components/webspace/settings/main.js]: back-button should be removed i will do it (:

File: src/Sulu/Bundle/ContentBundle/Resources/public/scss/modules/webspace-settings.scss (hunk 8 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/Resources/public/scss/modules/webspace-settings.scss]: yes in the husky variables!

File: src/Sulu/Bundle/SnippetBundle/Admin/WebspaceContentNavigationProvider.php (hunk 9 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/SnippetBundle/Admin/WebspaceContentNavigationProvider.php]: because i thought that the "main" bundle should set numbers factor 10 and the other bundles can use numbers in between! otherwise contentbundle has to know that snippet bundle uses number 30

File: src/Sulu/Bundle/SnippetBundle/Controller/SnippetTypesController.php (hunk 12 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/SnippetBundle/Controller/SnippetTypesController.php]: i have renamed it! but i will us the `@Route`.

File: src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php (hunk 4 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php]: moved the logic of filtering by webspacekey there.

File: src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php (hunk 5 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/ContentBundle/EventListener/WebspaceSerializeEventSubscriber.php]: i can use the url which is also unique in this context.

File: src/Sulu/Bundle/SnippetBundle/Controller/SnippettypesController.php (hunk 10 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/SnippetBundle/Controller/SnippettypesController.php]: fixed

File: src/Sulu/Bundle/SnippetBundle/Controller/SnippettypesController.php (hunk 11 above)
Comment (wachterjohannes) [inline src/Sulu/Bundle/SnippetBundle/Controller/SnippettypesController.php]: fixed

File: src/Sulu/Component/Webspace/Settings/SettingsManagerInterface.php (hunk 18 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Webspace/Settings/SettingsManagerInterface.php]: fixed

File: src/Sulu/Component/Webspace/Manager/WebspaceManager.php [hunk 19]
```diff
@@ -124,6 +125,35 @@ public function findUrlsByResourceLocator(
...
+        foreach ($portals as $url => $portalInformation) {
+            $sameLocalization = $portalInformation->getLocalization()->getLocalization() === $languageCode;
+            $sameWebspace = $webspaceKey === null || $portalInformation->getWebspace()->getKey() === $webspaceKey;
+            $url = rtrim(sprintf('%s://%s%s', $scheme, $url, $resourceLocator), '/');
+            if ($sameLocalization && $sameWebspace && $this->isFromDomain($url, $domain)) {
+                if ($portalInformation->isMain()) {
+                    array_unshift($urls, $url);
+                } else {
+                    $urls[] = $url;
+                }
```
Comment (wachterjohannes) [inline src/Sulu/Component/Webspace/Manager/WebspaceManager.php]: what do you think about this logic? @chirimoya 

It returns a main url it there is one for this domain and locale otherwise it returns the first url which is found from this domain.

**Comment by wachterjohannes** [comment]
@danrot we cannot rebase this PR because of the depending PR https://github.com/sulu-io/sulu/pull/1875 we have to solve the conflicts when merging.

File: UPGRADE.md [hunk 20]
```diff
@@ -1,5 +1,25 @@
...
+
+A new phpcr namespace was added. To register it run following command:
+
+```bash
+app/console sulu:phpcr:init
+```
+
+### ContentNavigation & Navigation
+
+The ContentNavigationItems & NavigationItems will be sorted by their position. If there is no position is set the item
```
Comment (danrot) [inline UPGRADE.md]: `If there is no position set, the item `

File: src/Sulu/Bundle/SnippetBundle/Snippet/SnippetResolverInterface.php [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace Sulu\Bundle\SnippetBundle\Snippet;
+
+/**
+ * Resolves snippets by UUIDs.
+ */
+interface SnippetResolverInterface
+{
+    /**
+     * Load snippet and reolves them by UUID.
```
Comment (danrot) [inline src/Sulu/Bundle/SnippetBundle/Snippet/SnippetResolverInterface.php]: resolves

File: src/Sulu/Bundle/SnippetBundle/Snippet/WrongSnippetTypeException.php [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+
+    /**
+     * @var SnippetDocument
+     */
+    private $document;
+
+    public function __construct($actual, $expected, SnippetDocument $document)
+    {
+        parent::__construct(
+            sprintf('Wrong snippet type were detected (actual: "", expected: "").', $actual, $expected)
```
Comment (danrot) [inline src/Sulu/Bundle/SnippetBundle/Snippet/WrongSnippetTypeException.php]: two `%s` missing

File: src/Sulu/Component/Webspace/Manager/WebspaceManager.php [hunk 23]
```diff
@@ -124,6 +125,35 @@ public function findUrlsByResourceLocator(
     /**
      * {@inheritdoc}
      */
+    public function findUrlByResourceLocator(
+        $resourceLocator,
+        $environment,
+        $languageCode,
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Manager/WebspaceManager.php]: Shouldn't that be called `locale`?

File: src/Sulu/Component/Webspace/Manager/WebspaceManagerInterface.php [hunk 24]
```diff
@@ -96,6 +117,15 @@ public function getUrls($environment);
     public function getPortalInformations($environment);
 
     /**
+     * Returns the portal informations for given webspace managed by this WebspaceManger.
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Manager/WebspaceManagerInterface.php]: `for the given webspace` and `by this WebspaceManager`

File: src/Sulu/Component/Webspace/Manager/WebspaceManager.php (hunk 23 above)
Comment (danrot) [inline src/Sulu/Component/Webspace/Manager/WebspaceManager.php]: I leave it as it is, because this way it is consistent.
